samples/echo/src/main/cpp/AudioEffect.cpp

```diff
@@ -0,0 +1,277 @@
+/*
+ * Copyright 2018 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "AudioEffect.h"
+#include <logging_macros.h>
+#include <climits>
+#include <cstring>
+#include <audio_common.h>
+
+/*
+ * Mixing Audio in integer domain to avoid FP calculation
+ *   (FG * ( MixFactor * 16 ) + BG * ( (1.0f-MixFactor) * 16 )) / 16
+ */
+static const int32_t kFloatToIntMapFactor = 128;
+
+AudioMixer::AudioMixer() :
+    AudioFormat(48000, 2, oboe::AudioFormat::I16) {
+
+  busy_ = false;
+  bgMixFactorInt_ = (int32_t)
+                   (bgMixFactor_ * kFloatToIntMapFactor + 0.5f);
+  fgMixFactorInt_ = kFloatToIntMapFactor - bgMixFactorInt_;
+}
+
+/**
+ * destructor: release memory for audio samples
+ */
+AudioMixer::~AudioMixer() {
+}
+/**
+ * Set mix factor for the 2 streams( background and forground );
+ * blending:
+ *    recordedAudio * fgMix +
+ *    backgroundMusic * ( 1.0f - fgMix )
+ * @param fgMix is background music mixer
+ */
+void  AudioMixer::setBackgroundMixer(float mixer) {
+  if (mixer >= 0.0f && mixer <= 1.0f) {
+    bgMixFactor_ = mixer;
+    bgMixFactorInt_ = (int32_t)
+      (bgMixFactor_ * kFloatToIntMapFactor + 0.5f);
+    fgMixFactorInt_ = kFloatToIntMapFactor - bgMixFactorInt_;
+  }
+}
+
+/**
+ * Insert a raw PCM audio buffer to blend with live audio
+ *
+ * @param samples points to PCM audio buffer
+ * @param sampleCount is total samples pointed by samples
+ * @param channelCount channels for PCM audio pointed by samples
+ * @param freq is PCM audio frequency (48000hz for this sample)
+ */
+void  AudioMixer::addStream(std::unique_ptr<int16_t[]>samples, size_t sampleCount,
+          int32_t sampleRate, int32_t channelCount, oboe::AudioFormat format){
+  if (busy_) {
```

Comment:
What would happen if `busy_` flips to `true` right after this check?

Reply:
application needs to retry the command. I will change it to return bool to  indicate the status. 
otherwise have to add a mutex ( it is not that bad either, the delay is using that method ) to let process() try and lock; on addStream() will blocking lock.

let me know for other ways I could borrow

Reply:
I see what you mean: yeah there is a brief window of contention. I will add lock to it